### PR TITLE
Nested folders

### PR DESCRIPTION
**Pull request by AELSchauer**
I got the nested folders functionality complete, including model tests 👍 

The two methods include finding the ancestors for a sub-folder (which we'll need for breadcrumbing on our views), and returning the root folder for a user.

I also added validations to the user and folder since I was writing the model tests.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -28,15 +28,20 @@ gem 'jbuilder', '~> 2.5'
...
 group :development, :test do
   # Call 'byebug' anywhere in the code to stop execution and get a debugger console
   gem 'byebug', platform: :mri
+  gem 'pry'
   gem 'rspec-rails'
+  gem 'shoulda-matchers'
+  gem 'factory_girl_rails'
+  gem 'faker'
+  gem 'capybara'
 end
```
Comment (cmacaulay) [inline Gemfile]: 🎉

File: app/models/folder.rb [hunk 2]
```diff
@@ -1,2 +1,19 @@
 class Folder < ApplicationRecord
+  belongs_to :user
+  belongs_to :parent, :class_name => 'Folder', optional: true
+  has_many :children, :class_name => 'Folder', :foreign_key => 'parent_id'
```
Comment (cmacaulay) [inline app/models/folder.rb]: 🎉 awesome

File: app/models/user.rb [hunk 3]
```diff
@@ -1,2 +1,15 @@
...
+
+  has_many :folders
+  has_many :uploads, through: :folders
+
+  validates :username,        presence: true, uniqueness: true
+  validates :cellphone,       presence: true, uniqueness: true
+  validates :email,           presence: true, uniqueness: true
+  validates :password_digest, presence: true
+
+  def root_folder
```
Comment (cmacaulay) [inline app/models/user.rb]: 🎉 @AELSchauer @JonathanSR @alfosco let's remember to add flash messages to the log-in/account creation around these validations.

File: spec/factories/folders.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+FactoryGirl.define do
+  factory :folder do
+    sequence(:name) { |n| "Example name #{n}" }
+    parent nil
+    user
+  end
+end
```
Comment (cmacaulay) [inline spec/factories/folders.rb]: @AELSchauer thank you for setting this up! 🎉

File: spec/factories/users.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
+FactoryGirl.define do
+  user_name = Faker::Internet.user_name
```
Comment (cmacaulay) [inline spec/factories/users.rb]: @AELSchauer please remove this line 😄

File: spec/models/folder_spec.rb [hunk 6]
```diff
@@ -1,5 +1,23 @@
...
+      user  = create(:user)
+      root  = create(:folder, user: user)
+      one   = create(:folder, user: user, parent: root)
+      two   = create(:folder, user: user, parent: one)
+      three = create(:folder, user: user, parent: two)
+
+      expect(three.ancestors).to eq([three, two, one, root])
+    end
+  end
 end
```
Comment (cmacaulay) [inline spec/models/folder_spec.rb]: 🎉 🎉

File: spec/models/user_spec.rb [hunk 7]
```diff
@@ -1,5 +1,22 @@
...
+
+  describe '#root_folder' do
+    it 'should return the root_folder for a user' do
+      user = create(:user)
+      root = create(:folder, user: user)
+
+      expect(user.root_folder).to eq(root)
+    end
+  end
 end
```
Comment (cmacaulay) [inline spec/models/user_spec.rb]: 🎉

File: spec/rails_helper.rb [hunk 8]
```diff
@@ -35,6 +37,13 @@
...
 
+  Shoulda::Matchers.configure do |config_shoulda|
+    config_shoulda.integrate do |with|
+      with.test_framework :rspec
+      with.library :rails
+    end
+  end
+
   # RSpec Rails can automatically mix in different behaviours to your tests
   # based on their file location, for example enabling you to call `get` and
```
Comment (cmacaulay) [inline spec/rails_helper.rb]: @AELSchauer really appreciate you setting all of this up.

**Comment by AELSchauer** [comment]
@cmacaulay I've fixed that typo line in the user factory. 👍

File: app/models/folder.rb (hunk 2 above)
Comment (alfosco) [inline app/models/folder.rb]: 👍

File: app/models/user.rb (hunk 3 above)
Comment (alfosco) [inline app/models/user.rb]: The 'username' column for 'users' table has been removed and replaced by 'first_name' and 'last_name'.

**Pull request merged**